cpp/ql/src/semmle/code/cpp/internal/AddressConstantExpression.qll

```diff
@@ -0,0 +1,191 @@
+private import cpp
+private import semmle.code.cpp.dataflow.EscapesTree
+
+predicate addressConstantExpression(Expr e) {
+  constantAddressPointer(e)
+  or
+  constantAddressReference(e)
+  or
+  // Special case for function pointers, where `fp == *fp`.
+  constantAddressLValue(e) and
+  e.getType() instanceof FunctionPointerType
+}
+
+/** Holds if `v` is a constexpr variable initialized to a constant address. */
+private predicate addressConstantVariable(Variable v) {
+  addressConstantExpression(v.getInitializer().getExpr().getFullyConverted()) and
+  // Here we should also require that `v` is constexpr, but we don't have that
+  // information in the db. See CPP-314. Instead, we require that the variable
+  // is not assigned to.
+  not exists(VariableAccess va | va.getTarget() = v |
+    // `v` may be assigned to, completely or partially
+    exists(Expr lvalue | variableAccessedAsValue(va, lvalue) |
+      lvalue = any(Assignment a).getLValue().getFullyConverted()
+      or
+      lvalue = any(CrementOperation c).getOperand().getFullyConverted()
+    )
+  )
+}
+
+/**
+ * Holds if `lvalue` is an lvalue whose address is an _address constant
+ * expression_.
+ */
+private predicate constantAddressLValue(Expr lvalue) {
+  lvalue.(VariableAccess).getTarget() = any(Variable v |
+      v.(Variable).isStatic()
+      or
+      v instanceof GlobalOrNamespaceVariable
+    )
+  or
```

Comment:
What about members of global or static variables?
```
global.x = ...
```

Reply:
These are addressed in the recursive cases: https://github.com/Semmle/ql/blob/7c44764aff43fd1a928a6e08b5e7696a5b612cbc/cpp/ql/src/semmle/code/cpp/internal/AddressConstantExpression.qll#L113